libraries/model-networking/src/model-networking/ModelCache.cpp

```diff
@@ -55,458 +103,350 @@ void GeometryReader::run() {
     }
 
     QThread::currentThread()->setPriority(originalPriority);
 }
 
-NetworkGeometry::NetworkGeometry(const QUrl& url, bool delayLoad, const QVariantHash& mapping, const QUrl& textureBaseUrl) :
-    _url(url),
-    _mapping(mapping),
-    _textureBaseUrl(textureBaseUrl.isValid() ? textureBaseUrl : url) {
+class GeometryDefinitionResource : public GeometryResource {
+    Q_OBJECT
+public:
+    GeometryDefinitionResource(const QUrl& url, const QVariantHash& mapping, const QUrl& textureBaseUrl) :
+        GeometryResource(url), _mapping(mapping), _textureBaseUrl(textureBaseUrl) {}
 
-    if (delayLoad) {
-        _state = DelayState;
-    } else {
-        attemptRequestInternal();
-    }
+    virtual void downloadFinished(const QByteArray& data) override;
+
+protected:
+    Q_INVOKABLE void setGeometryDefinition(void* fbxGeometry);
+
+private:
+    QVariantHash _mapping;
+    QUrl _textureBaseUrl;
+};
+
+void GeometryDefinitionResource::downloadFinished(const QByteArray& data) {
+    QThreadPool::globalInstance()->start(new GeometryReader(_self, _url, _mapping, data));
 }
 
-NetworkGeometry::~NetworkGeometry() {
-    if (_resource) {
-        _resource->deleteLater();
+void GeometryDefinitionResource::setGeometryDefinition(void* fbxGeometry) {
+    // Assume ownership of the geometry pointer
+    _geometry.reset(static_cast<FBXGeometry*>(fbxGeometry));
+
+    // Copy materials
+    QHash<QString, size_t> materialIDAtlas;
+    for (const FBXMaterial& material : _geometry->materials) {
+        materialIDAtlas[material.materialID] = _materials.size();
+        _materials.push_back(std::make_shared<NetworkMaterial>(material, _textureBaseUrl));
     }
-}
 
-void NetworkGeometry::attemptRequest() {
-    if (_state == DelayState) {
-        attemptRequestInternal();
+    std::shared_ptr<NetworkMeshes> meshes = std::make_shared<NetworkMeshes>();
+    std::shared_ptr<NetworkShapes> shapes = std::make_shared<NetworkShapes>();
+    int meshID = 0;
+    for (const FBXMesh& mesh : _geometry->meshes) {
+        // Copy meshes
+        meshes->push_back(mesh._mesh);
+        int partID = 0;
+        for (const FBXMeshPart& part : mesh.parts) {
+            // Construct shapes
+            shapes->push_back(std::make_shared<NetworkShape>(meshID, partID, (int)materialIDAtlas[part.materialID]));
+            partID++;
+        }
+        meshID++;
     }
+    _meshes = meshes;
+    _shapes = shapes;
+
+    finishedLoading(true);
 }
 
-void NetworkGeometry::attemptRequestInternal() {
-    if (_url.path().toLower().endsWith(".fst")) {
-        _mappingUrl = _url;
-        requestMapping(_url);
+ModelCache::ModelCache() {
+    const qint64 GEOMETRY_DEFAULT_UNUSED_MAX_SIZE = DEFAULT_UNUSED_MAX_SIZE;
+    setUnusedResourceCacheSize(GEOMETRY_DEFAULT_UNUSED_MAX_SIZE);
+}
+
+QSharedPointer<Resource> ModelCache::createResource(const QUrl& url, const QSharedPointer<Resource>& fallback,
```

Comment:
I wish this returned a raw pointer and wrapped it in `QSharedPointer` for you so you couldn't forget the special `&Resource::allReferencesCleared` dtor.